### ISSUE TITLE
Execution pattern changed with parameters in between input and output

### ISSUE THREAD

**Issue by gaow** [vatlab/sos#1375]
We ran into this problem that took us a while to figure out. It boils down to this MWE:

```
[global]
parameter: num = [x+1 for x in range(5)]

[1]
input: for_each = 'num'
output: f'{_num}.txt'
bash: expand = True
  touch {_output}

[2]
depends: executable('stress')
parameter: a = 1
output: f'{_input:n}.out'
bash: expand = True
  stress -c 1
```

Running this workflow I expect the 2nd step to execute in parallel. However it does not. Also when trying to terminate it using `ctrl-c` it just cannot terminal all the `stress` processes. One has to manually `kill` them, or reboot the computer when this step has hundreds of substeps that cannot be killed easily. 

However if we comment out `parameter: a = 1`, it will run in parallel and `ctrl-c` kill will work. The message reported in response to `ctrl-c` is also different.

For now, we have to explicitly specify `input` to get it work as desired:

```
[2]
depends: executable('stress')
parameter: a = 1
input: group_by = 1
output: f'{_input:n}.out'
bash: expand = True
  stress -c 1
```

The actual workflow we have is a bit more complicated, like:

```
parameter: a = 1
if ... : 
 ...
else: 
 ...
```

and these parameters are step specific to be used with nested workflow, so we cannot readily move them to `global`. But still I think the behavior above, though might be justifiable from a design prospective, should be considered a bug from user's point of view.

**Comment by BoPeng** [comment]
I will have a look.

DockerCon is going on now, get on if you have some time.

**Comment by BoPeng** [comment]
Maybe this is related to my recent work on local parameters #1372 , which should not have been released. Are you on master or latest release?

**Comment by gaow** [comment]
Thanks @BoPeng !

We are on version 0.21.8 which should have incorporated that patch.

**Comment by BoPeng** [comment]
For the concurrent execution part, 

```
✗ sos run test -v3
INFO: Running 1:
INFO: 1 (index=0) is ignored due to saved signature
INFO: 1 (index=1) is ignored due to saved signature
INFO: 1 (index=2) is ignored due to saved signature
INFO: 1 (index=3) is ignored due to saved signature
INFO: 1 (index=4) is ignored due to saved signature
INFO: 1 output:   1.txt 2.txt... (5 items in 5 groups)
INFO: Running 2:
DEBUG: Substeps are executed sequentially because of existence of directives between statements.
```

Basically, sos considers everything after `input` as part of substep. If substep starts with `output:`, `depends:`, they will be interpreted before substeps are sent. However, if it sees a regular statement, in this case a parameter assignment, it cannot execute the substep in parallel because it has to execute part of the substep in the master step, and process the `output:` statement. Moving `parameter:` after `output:` fixes the problem.

**Comment by gaow** [comment]
So the general recommendation is for each `step` we have :+1: 
```
depends:
input:
output:
parameter 
```

or, 

```
input:
output:
depends:
parameter 
```

? 

currently I specify input and output after depends and parameter. We can certainly change the ordering to get it work. It is just that the behavior is somewhat unexpected. 

Also did you notice the different behavior when you try to `ctrl-c` to terminate? When we have `parameter` before `output`, it does not terminate properly.

**Comment by gaow** [comment]
> . Moving parameter: after output: fixes the problem.

What if  `output` name depends on `parameter` value?

**Comment by BoPeng** [comment]
> What if output name depends on parameter value?

That was the issue, also will be seen in the case when the output depends on some other expression derived from `input`. 

I think this issue is fixable if we change the definition of "substep" to be things after all statements (e.g. `input`, `output`, `depends`) and execute everything in between in the master process. I do not know how difficult such change can be though. The tricky part is that users might think `parameter:` is a statement, but in implementation it is translated to a regular function.

**Comment by BoPeng** [comment]
Interesting, the current implementation does treat things after `output` as substep, so the restriction, which was probably introduced earlier, is not necessary.

```
[global]
import os
parameter: num = [x+1 for x in range(5)]

[1]
input: for_each = 'num'
output: f'{_num}.txt'
bash: expand = True
  touch {_output}

[2]
depends: executable('stress')
parameter: a = 1
print(f'I am in between {os.getpid()}')
output: f'{_input:n}.out'
print(f'I am after {os.getpid()}')
bash: expand = True
  stress -c 1
```
Output
```
(sos) ➜  t1375 git:(master) ✗ sos run test -s force
INFO: Running 1:
INFO: 1 (index=0) is completed.
INFO: 1 (index=1) is completed.
INFO: 1 (index=2) is completed.
INFO: 1 (index=3) is completed.
INFO: 1 (index=4) is completed.
INFO: 1 output:   1.txt 2.txt... (5 items in 5 groups)
INFO: Running 2:
I am in between 86519
I am in between 86519
I am in between 86519
I am in between 86519
I am in between 86519
I am after 86514
stress: info: [86535] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd
I am after 86517
stress: info: [86538] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd
I am after 86512
I am after 86520
I am after 86511
stress: info: [86542] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd
stress: info: [86544] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd
stress: info: [86547] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd
```

**Comment by gaow** [comment]
Sorry I dont understand -- did you remove the restriction in your local copy of the code and made this work? What was the reason to make the restriction (perhaps `git blame` can help)?

**Comment by gaow** [comment]
Oh okay just saw this patch. It's quite clear it was made to restrict derivatives between statements. However I'm still wondering why I was not able to kill the job with `ctrl-c`. We had to reboot the computer a couple of times today to terminate a few hundred substeps ...

**Comment by BoPeng** [comment]
When `Ctrl-C` kills a process, it will kill all its subprocesses (usually). In this case `Ctrl-C` is intercepted and stopped the substep, and allows the next substep to be executed in the same worker process. Then I think when you forcefully kills SoS without letting it to clean up, the subprocesses will be kept unkilled.

**Comment by gaow** [comment]
So in the case with `concurrent=False` we basically cannot easily kill all substeps because they will keep popping up? concurrent = False might be triggered in nested workflow. The behavior doesn't seem very friendly for that case.

**Comment by BoPeng** [comment]
Let me investigate this a bit more... In the case of `concurrent=False` only one worker is processing all the substeps, but `Ctrl-C` is still sent to all the workers. It has been tricky to handle `Ctrl-C` especially when `zmq` is involved because `sos` will freeze if one of the `zmq`-enabled worker is killed and fails to answer zmq messages.

**Comment by gaow** [comment]
The `ctrl-c` issue is now a separate ticket #1377 . For this ticket there is yet still an issue **on the cluster**. Here is the MWE:

```
[global]
parameter: num = [x+1 for x in range(5)]

[1]
input: for_each = 'num'
output: f'{_num}.txt'
bash: expand = True
  touch {_output}

[2]
parameter: a = 1
output: f'{_input:n}.out'
task: trunk_workers = 1, trunk_size = 1, walltime = '3m', mem = '1G', cores = 1
bash: expand = True
  echo {a} > {_output}
```

Running it locally:

```
sos run test.sos
```

it works well. Running it as tasks on the cluster:

```
MW] sos run test.sos -c midway2.yml -q midway2
INFO: Running 1: 
INFO: 1 (index=0) is ignored due to saved signature
INFO: 1 (index=1) is ignored due to saved signature
INFO: 1 (index=2) is ignored due to saved signature
INFO: 1 (index=3) is ignored due to saved signature
INFO: 1 (index=4) is ignored due to saved signature
INFO: 1 output:   1.txt 2.txt... (5 items in 5 groups)
INFO: Running 2: 
ERROR: [2]: [2]: Output target 1.out does not exist after the completion of step 2
```
it does not even submit any jobs. But after commenting out `parameter: a = 1` it works fine. Hopefully it is not a hard fix!

**Comment by BoPeng** [comment]
Consequence of redefining "substep". Please try again.

**Comment by gaow** [comment]
Still doesn't work, due to another recent bug I suppose:

```
[global]
parameter: num = [x+1 for x in range(5)]

[1]
input: for_each = 'num'
output: f'{_num}.txt'
bash: expand = True
  touch {_output}

[2]
parameter: a = 1
output: f'{_input:n}.out'
sos_run('a' if a >=1 else 'b')

[a,b]
output: f'{_input:n}.out'
bash: expand = True
  touch {_output}
```

```
[MW] sos run test.sos 
INFO: Running 1: 
INFO: 1 (index=3) is completed.
INFO: 1 (index=0) is completed.
INFO: 1 (index=4) is completed.
INFO: 1 (index=2) is completed.
INFO: 1 (index=1) is completed.
INFO: 1 output:   1.txt 2.txt... (5 items in 5 groups)
INFO: Running 2: 
INFO: 2 (index=0) is completed (pending nested workflow).
INFO: Running a: 
INFO: a is completed.
INFO: a output:   1.out
ERROR: [2]: [2]: '>' not supported between instances of 'NoneType' and 'int'
```

**Comment by BoPeng** [comment]
Sorry, I did not realized that `-j` is now very complex and is not processed upfront. Try again.

**Comment by gaow** [comment]
Yes it works now. Thanks a lot @BoPeng !
